internal/access/grant.go

```diff
@@ -18,33 +22,49 @@ func GetGrant(c *gin.Context, id uid.ID) (*models.Grant, error) {
 }
 
 func ListGrants(c *gin.Context, subject uid.PolymorphicID, resource string, privilege string) ([]models.Grant, error) {
-	db, err := RequireInfraRole(c, models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole)
-	if err != nil {
-		return nil, err
+	selectors := []data.SelectorFunc{
+		data.ByOptionalResource(resource),
+		data.ByOptionalPrivilege(privilege),
 	}
 
-	return data.ListGrants(db,
-		data.ByOptionalSubject(subject),
-		data.ByOptionalResource(resource),
-		data.ByOptionalPrivilege(privilege))
-}
+	db, err := RequireInfraRole(c, models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole)
+	if err == nil {
+		selectors = append(selectors, data.ByOptionalSubject(subject))
+		return data.ListGrants(db, selectors...)
+	}
 
-func ListIdentityGrants(c *gin.Context, identityID uid.ID) ([]models.Grant, error) {
-	db, err := hasAuthorization(c, identityID, isIdentitySelf, models.InfraAdminRole, models.InfraViewRole)
-	if err != nil {
-		return nil, err
+	if errors.Is(err, internal.ErrForbidden) {
+		// Allow an authenticated identity to view their own grants
+		db := getDB(c)
+		subjectID, _ := subject.ID()
+		identity := AuthenticatedIdentity(c)
+		switch {
+		case identity == nil:
+			return nil, err
+		case subject.IsIdentity() && identity.ID == subjectID:
+			selectors = append(selectors, data.ByOptionalSubject(subject))
```

Comment:
The subject should not be optional in this case should it? The caller without permissions should only be able to see their own resources.

Reply:
Yes, good catch! I fixed exactly this bug in an earlier commit, but then accidentally restored it when I went to add the other selectors to this query.  It should be `BySubject`.

I still wish we did not have `Optional` selectors, and just used an if statement.